### PR TITLE
fix: Unhandled exception when calling `Parse.Cloud.run` with option value `null` (#2622)

### LINKED ISSUE #2622: Error when passing `null` as `Parse.Cloud.run` option
### New Issue Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [x] I have searched through [existing issues](https://github.com/parse-community/Parse-SDK-JS/issues?q=is%3Aissue).
- [x] I can reproduce the issue with the latest versions of [Parse Server](https://github.com/parse-community/parse-server/releases) and the [Parse JS SDK](https://github.com/parse-community/Parse-SDK-JS/releases). 

### Issue Description

This bug was first found in https://github.com/parse-community/parse-server/ (see issue https://github.com/parse-community/parse-server/issues/9766)

In `_getRequestOptions` in file `ParseObject.js` the parameter `options` is set to `{}` if it has no value. This works for values like `''`, `undefined` or if it is empty, but if it is `null` the code breaks.

### Steps to reproduce

Call the function `Parse.Cloud.run(<name>, data, null);` or `_ParseObject.default._getRequestOptions(null);` directly.

### Actual Outcome

Error:

```
TypeError: Cannot convert undefined or null to object
    at hasOwn (<anonymous>)
    at Function._getRequestOptions (/Users/#################/node_modules/parse/lib/node/ParseObject.js:423:9)
    at Object.run (/Users/#################/node_modules/parse/lib/node/Cloud.js:57:47)
…

```
### Expected Outcome

When no options `_getRequestOptions` should return `{}`

Some kind of null check for the parameter `options`.

```
if(!options){
    return requestOptions;
}
```

### Environment

I have tested this in `8.2.0` and `8.2.1-alpha.2`. Error occurs in both.

The function does not exists in `7.5.2`

Server
- Parse Server version: `8.2.0`
- Operating system: `macOS 15.5`
- Local or remote host (AWS, Azure, Google Cloud, Heroku, Digital Ocean, etc): `local`

Database
- System (MongoDB or Postgres): `MongoDB`
- Database version: `macOS 15.5`
- Local or remote host (MongoDB Atlas, mLab, AWS, Azure, Google Cloud, etc): `local`

Client
- SDK (iOS, Android, JavaScript, PHP, Unity, etc): `Node`
- SDK version: `22.14.0`

Client
- Parse JS SDK version: `6.1.0`

### LINKED ISSUE #9766: Error when passing `null` as `Parse.Cloud.run` option
### New Issue Checklist

- Report security issues [confidentially](https://github.com/parse-community/parse-server/security/policy).
- Any contribution is under this [license](https://github.com/parse-community/parse-server/blob/alpha/LICENSE).
- Before posting search [existing issues](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Issue Description

In `_getRequestOptions` in file `ParseObject.js` the parameter `options` is set to `{}` if it has no value. This works for values like `''`, `undefined` or if it is empty, but if it is `null` the code breaks.

### Steps to reproduce

Call the function `Parse.Cloud.run(<name>, data, null);` or `_ParseObject.default._getRequestOptions(null);` directly.

### Actual Outcome

Error:

```
TypeError: Cannot convert undefined or null to object
    at hasOwn (<anonymous>)
    at Function._getRequestOptions (/Users/#################/node_modules/parse/lib/node/ParseObject.js:423:9)
    at Object.run (/Users/#################/node_modules/parse/lib/node/Cloud.js:57:47)
…

```
### Expected Outcome

When no options `_getRequestOptions` should return `{}`

### Environment

I have tested this in `8.2.0` and `8.2.1-alpha.2`. Error occurs in both.

The function does not exists in `7.5.2`

Server
- Parse Server version: `8.2.0`
- Operating system: `macOS 15.5`
- Local or remote host (AWS, Azure, Google Cloud, Heroku, Digital Ocean, etc): `local`

Database
- System (MongoDB or Postgres): `MongoDB`
- Database version: `macOS 15.5`
- Local or remote host (MongoDB Atlas, mLab, AWS, Azure, Google Cloud, etc): `local`

Client
- SDK (iOS, Android, JavaScript, PHP, Unity, etc): `Node`
- SDK version: `22.14.0`

### Proposed solution
Some kind of null check for the parameter `options`. Like insert this on row 420:

```
if(!options){
    return requestOptions;
}
```

### PR DESCRIPTION
**Pull request by modime**
## Pull Request

- Report security issues [confidentially](https://github.com/parse-community/Parse-SDK-JS/security/policy).
- Any contribution is under this [license](https://github.com/parse-community/Parse-SDK-JS/blob/alpha/LICENSE).
- Link this pull request to an [issue](https://github.com/parse-community/Parse-SDK-JS/issues?q=is%3Aissue).

This Pull Request is a fix for when passing null as a Parse.Cloud.run option 

## Issue

Closes: https://github.com/parse-community/Parse-SDK-JS/issues/2622
Closes: https://github.com/parse-community/parse-server/issues/9766

## Tasks

- [x] Add tests
- [ ] Add changes to documentation (guides, repository pages, code comments)

## Summary by CodeRabbit

- **Bug Fixes**
  - Enhanced validation to accept only plain objects for the options parameter, preventing invalid input types.

- **Tests**
  - Added tests to ensure proper handling of options, including acceptance of empty or undefined values and rejection of invalid types.

### CODE REVIEW

**Comment by parseplatformorg** [comment]
### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/acinader/pr-checks/3e5eaa73-7fb5-44ae-911f-792a434d48ca?refs=29fa2c53c4272a0007f154ab4ab54477ea010b82&source=prChecksComment)

File: src/__tests__/ParseObject-test.js [hunk 1]
```diff
@@ -3543,3 +3543,23 @@ describe('ParseObject pin', () => {
     CoreManager.set('NODE_LOGGING', false);
   });
 });
+
+describe('ParseObject._getRequestOptions', () => {
```
Comment (mtrezza) [inline src/__tests__/ParseObject-test.js]: Well done, you have also added tests! We'd add the to a higher API like `Parse.Cloud.run` which invokes `_getRequestOptions`, rather than on `_getRequestOptions` itself. The test would be added to `/__tests__/Cloud-test.js`. You can also compact into a single "empty options" test:

```js
it('run passes with empty options', () => {
  const values = [
    undefined,
    {},
  ];    

  for (const value of values) {
    expect(Cloud.run('myfunction', {}, value)).not.toThrow();
    expect(CoreManager.getCloudController().run.mock.calls[0]).toEqual([
      'myfunction',
      {},
      {},
    ]);
  });
```

Note that I've deliberately left out the `null` and empty string value, see my other feedback. We should then also add a test where we expect to throw on invalid values like `null`, empty string, `false`, array, etc.

File: src/ParseObject.ts [hunk 2]
```diff
@@ -514,6 +514,9 @@ class ParseObject<T extends Attributes = Attributes> {
 
   static _getRequestOptions(options: RequestOptions & FullOptions & { json?: boolean } = {}) {
     const requestOptions: RequestOptions & FullOptions & { json?: boolean } = {};
+    if (!options) {
```
Comment (mtrezza) [inline src/ParseObject.ts]: We could skip on undefined and throw a nice error message on invalid value:

```suggestion
    if (options === undefined) {
      return requestOptions;
    }
    if (Object.prototype.toString.call(options) !== '[object Object]') {
      throw new Error('request options must be of type Object');
    }
```

**Review by mtrezza — COMMENTED** [review]
Thanks for this PR! The first question is, do we really want a call to silently succeed, even if the developer made a coding mistake?

- Passing a string while an object is expected is a type violation, ergo clearly a coding mistake to which we want to alert the developer, so throwing makes sense.
- Passing `null` may or may not be considered a coding mistake, but at a quick glance it seems like one to me. The allowed type for the `options` argument is `undefined` (no options set), or an object. Anything else is a developer error. A variable that holds `null` means that the developer deliberately assigned no value to it, without ensuring that it is assigned the value of an object before passing it to the method. Hence it's a type violation as well.

**Comment by mtrezza** [comment]
@coderabbitai review

**Comment by modime** [comment]
@mtrezza Thanks for the feedback. Update this MR as you see fit. My take on this was mostly a bug fix: If "no options" then return an empty bracket, i.e. handle `null` the same as `undefined` (and also return early in the code when there are no options to act on).

We found this error in `Parse.Cloud.run` (parse-server 7.5.2) when our tests with options set to `null` expects an `Invalid request.` error with the error code `141`.

**Comment by mtrezza** [comment]
I believe when throwing like suggested it should return a proper error, instead of an internal server error. We'd have to see in the test. Would you want to incorporate the changes?

**Comment by modime** [comment]
> I believe when throwing like suggested it should return a proper error, instead of an internal server error. We'd have to see in the test. Would you want to incorporate the changes?

No, better you do it. I am unable to make changes for the next week as I am traveling and have limited access to the source code.

**Comment by modime** [comment]
@mtrezza I'm now back at work and I have now updated the code as I hope you wanted. Please review. Thanks.

File: src/ParseObject.ts (hunk 2 above)
Comment (modime) [reply src/ParseObject.ts]: As I see it the following part has unreachable code so I didn't include it. If `options` is `undefined` it will be set to `{}` by the default function parameter.

```
if (options === undefined) {
      return requestOptions;
}
```

**Comment by dplewis** [comment]
LGTM! I forgot the default option `{}` is only used for `undefined` not `null` as `null` is a value. Good find!

**Review by mtrezza — DISMISSED** [review]

File: src/__tests__/Cloud-test.js [hunk 3]
```diff
@@ -356,4 +356,31 @@ describe('CloudController', () => {
     });
     expect(options.useMasterKey).toBe(false);
   });
+
```
Comment (mtrezza) [inline src/__tests__/Cloud-test.js]:
```suggestion

```

File: src/__tests__/Cloud-test.js [hunk 4]
```diff
@@ -356,4 +356,31 @@ describe('CloudController', () => {
...
+    }
+  });
+
+  it('run throws with invalid options', () => {
+    const values = [null, []];
+    for (const value of values) {
+      expect(() => Cloud.run('myfunction', {}, value)).toThrow();
+    }
+  });
+
```
Comment (mtrezza) [inline src/__tests__/Cloud-test.js]:
```suggestion

```

File: src/__tests__/Cloud-test.js [hunk 5]
```diff
@@ -356,4 +356,29 @@ describe('CloudController', () => {
     });
     expect(options.useMasterKey).toBe(false);
   });
+  it('run passes with empty options', () => {
```
Comment (mtrezza) [inline src/__tests__/Cloud-test.js:360]:
```suggestion

  it('run passes with empty options', () => {
```

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [7.0.0-alpha.1](https://github.com/parse-community/Parse-SDK-JS/releases/tag/7.0.0-alpha.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [7.0.0](https://github.com/parse-community/Parse-SDK-JS/releases/tag/7.0.0)
